### PR TITLE
Don't autobind gridkeys, add stronger support for factory and builder tags

### PR DESCRIPTION
**Pull request by thehobojoe**
### Changes

This PR completely removes the automatic hotkey assignment for grid keys, and also adds proper hotkey support to the new `builder` and `factory` tags on grid hotkey assignment.

The purpose of this is to fix a couple separate problems.
- If a user wants to use specific hotkeys that are claimed by grid keys, in the current behavior they simply could not, because the script would take over and reassign them, resulting in weird double binds
- If a user replaced a generic gridmenu_key hotkey with one with a `builder` or `factory` tab e.g.
```
bind 		Any+sc_a  gridmenu_key 2 1 builder
bind           	Any+sc_s  gridmenu_key 2 2 builder
bind           	Any+sc_d  gridmenu_key 2 3 builder
bind           	Any+sc_w  gridmenu_key 3 2 builder
```
Then it simply would not work

### Shortcomings
- The arrangement from this PR means separate assignment for `builder` and `factory` tags is not possible, e.g. assigning `w` to `gridmenu_key 3 2 builder` and `meta+w` to `gridmenu_key 3 2 factory`. Since this already wasn't possible I think that's fine, this is something I would like to add in the future as I get more familiar with this lua code.
- If you have only `builder` tags assigned to a key (e.g. WASD), the hotkey hints will still show up in factories even though they do nothing. Again this is something I would like to add support for in the future when I know this lua better.

### CODE REVIEW

**Comment by thehobojoe** [comment]
@badosu would love a review on this

File: luaui/Widgets/gui_gridmenu.lua [hunk 1]
```diff
@@ -45,26 +45,6 @@ local paginatorFontSize, paginatorCellHeight
 local Cfgs = {
 	NEXT_PAGE_KEY = "SC_B",
```
Comment (badosu) [inline luaui/Widgets/gui_gridmenu.lua:46]: You can remove this one as well.

File: luaui/Widgets/gui_gridmenu.lua [hunk 2]
```diff
@@ -680,46 +660,53 @@ local function getActionHotkey(action)
...
 	if not key then
-		key = Cfgs.NEXT_PAGE_KEY
-		Spring.SendCommands('bind ' .. key .. ' gridmenu_next_page')
+		Spring.Echo("Error, missing grid key bind for next page, things may not function as expected")
 	end
 
 	Cfgs.NEXT_PAGE_KEY = key
 
-	key = getActionHotkey('gridmenu_prev_page')
+	key = getGridKey('gridmenu_prev_page')
```
Comment (badosu) [inline luaui/Widgets/gui_gridmenu.lua:706]: For `gridmenu_prev_page` bindings the `builder` or `factory` argument is invalid, just use `getActionHotkey`.

**Comment by badosu** [comment]
Just make sure that both default and grid opti keybinding presets work with this.

Also, if you're changing stuff for wasd compatibility with gridmenu, please add the wasd grid keys as a new preset so users can pick it there.

Looks good to me, but I didn't test, if everything works as you expect so far (minus the adjustments you mentioned for factory/builder keys) you can merge it.

**Comment by badosu** [comment]
For both shortcomings you mentioned the fix is rather simple, but annoying.

Basically you'll want to map `keyLayout` and `vKeyLayout` cache variables and calculations to `keyLayout['builder'|'factory']` and `vKeyLayout['builder'|'factory']` scoped caches.

When calculating them you'll check for each category assignment and if not assigned then checking if the general binding (without builder or factory argument) assignment.

But you should skip it for the scope of this PR, let's do one step at a time.

**Comment by thehobojoe** [comment]
> Just make sure that both default and grid opti keybinding presets work with this.

Already done 👍 

> Also, if you're changing stuff for wasd compatibility with gridmenu, please add the wasd grid keys as a new preset so users can pick it there.

Yep, my plan is to make those presets once this is all merged and I can make sure everything works as expected in prod.


> For both shortcomings you mentioned the fix is rather simple, but annoying.
> 
> Basically you'll want to map `keyLayout` and `vKeyLayout` cache variables and calculations to `keyLayout['builder'|'factory']` and `vKeyLayout['builder'|'factory']` scoped caches.
> 
> When calculating them you'll check for each category assignment and if not assigned then checking if the general binding (without builder or factory argument) assignment.
> 
> But you should skip it for the scope of this PR, let's do one step at a time.

Thanks for the details on this, that will be helpful in the follow-up PR's I'd like to do to expand the capabilities here a little more. But agreed, for now this is a pretty small-scale bugfix to the PR that added the `builder` and `factory` tags.

**Comment by thehobojoe** [comment]
> @thehobojoe The code looks good to me, but I defer to you testing it works with default and grid keys.
> 
> Another test to consider is if gridmenu keys are not bound that gridmenu does not crash buildmenu and is minimally usable (with mouse).

Yep I have tested this with all the default configurations, as well as with configurations missing default grid binds. This changes nothing with the UI or click behavior so mouse is uneffected. I'm waiting for @resopmok to test this just to be extra sure nothing broke.

**Comment by nbusseneau** [comment]
I've tested a bit the PR by manually putting in `luaui/Widgets` both the modified `gui_gridmenu.lua` and the following `uikeys.txt` (provided by @thehobojoe, supposed to match default Grid Keys configuration with just WASD camera keys enabled on top):

```
unbindall          // clear the defaults
unbind enter chat  // clear the defaults

fakemeta  space

bind              enter  chat
bind                esc  select AllMap++_ClearSelection+
bind                esc  quitmessage
bind          Shift+esc  quitmenu
bind     Ctrl+Shift+esc  quitforce
bind      Alt+Shift+esc  reloadforce
bind         Any+escape  edit_escape
bind          Any+pause  pause
bind                esc  teamstatus_close
bind                esc  customgameinfo_close
bind                esc  buildmenu_pregame_deselect
bind           Any+sc_z  selectbox_same
bind          Any+space  selectbox_idle
bind          Any+shift  selectbox_all
bind           Any+ctrl  selectbox_deselect
bind            Any+alt  selectbox_mobile
bind          Any+space  selectloop
bind           Any+ctrl  selectloop_invert
bind          Any+shift  selectloop_add
bind               sc_z  gridmenu_category 1
bind               sc_x  gridmenu_category 2
bind               sc_c  gridmenu_category 3
bind               sc_v  gridmenu_category 4
bind         Shift+sc_z  gridmenu_category 1
bind         Shift+sc_x  gridmenu_category 2
bind         Shift+sc_c  gridmenu_category 3
bind         Shift+sc_v  gridmenu_category 4
bind           Any+sc_z  gridmenu_key 1 1
bind           Any+sc_x  gridmenu_key 1 2
bind           Any+sc_c  gridmenu_key 1 3
bind           Any+sc_v  gridmenu_key 1 4
bind           Any+sc_f  gridmenu_key 2 4
bind           Any+sc_q  gridmenu_key 3 1
bind           Any+sc_e  gridmenu_key 3 3
bind           Any+sc_r  gridmenu_key 3 4
bind               sc_b  gridmenu_next_page
bind               sc_n  gridmenu_prev_page
bind          Any+enter  chat
bind      Alt+ctrl+sc_a  chatswitchally
bind      Alt+ctrl+sc_s  chatswitchspec
bind            Any+tab  edit_complete
bind      Any+backspace  edit_backspace
bind         Any+delete  edit_delete
bind           Any+home  edit_home
bind           Alt+left  edit_home
bind            Any+end  edit_end
bind          Alt+right  edit_end
bind             Any+up  edit_prev_line
bind           Any+down  edit_next_line
bind           Any+left  edit_prev_char
bind          Any+right  edit_next_char
bind          Ctrl+left  edit_prev_word
bind         Ctrl+right  edit_next_word
bind          Any+enter  edit_return
bind             Ctrl+v  pastetext
bind           Alt+sc_=  increasespeed
bind           Alt+sc_-  decreasespeed
bind        Alt+numpad+  increasespeed
bind        Alt+numpad-  decreasespeed
bind               sc_[  buildfacing inc
bind         Shift+sc_[  buildfacing inc
bind               sc_]  buildfacing dec
bind         Shift+sc_]  buildfacing dec
bind           Alt+sc_z  buildspacing inc
bind     Shift+Alt+sc_z  buildspacing inc
bind           Alt+sc_x  buildspacing dec
bind     Shift+Alt+sc_x  buildspacing dec
bind               sc_b  onoff
bind         Shift+sc_b  onoff
bind          Ctrl+sc_b  selfd
bind    Ctrl+Shift+sc_b  selfd queued
bind               sc_e  reclaim
bind         Shift+sc_e  reclaim
bind               sc_f  fight
bind         Shift+sc_f  fight
bind           Alt+sc_f  forcestart
bind               sc_g  stopproduction
bind         Shift+sc_g  stopproduction
bind               sc_g  stop
bind         Shift+sc_g  stop
bind               sc_h  patrol
bind         Shift+sc_h  patrol
bind               sc_i  unit_stats
bind          Ctrl+sc_i  customgameinfo
bind               sc_j  loadunits
bind         Shift+sc_j  loadunits
bind               sc_k  cloak
bind         Shift+sc_k  cloak
bind               sc_k  wantcloak
bind           Any+sc_k  wantcloak
bind               sc_l  firestate
bind         Shift+sc_l  firestate
bind               sc_;  movestate
bind         Shift+sc_;  movestate
bind               sc_m  restore
bind         Shift+sc_m  restore
bind               sc_n  settargetnoground
bind               sc_o  guard
bind         Shift+sc_o  guard
bind           Alt+sc_o  cameraflip
bind               sc_r  repair
bind         Shift+sc_r  repair
bind               sc_t  repeat
bind         Shift+sc_t  repeat
bind               sc_u  unloadunits
bind         Shift+sc_u  unloadunits
bind               sc_y  wait
bind         Shift+sc_y  wait queued
bind          Ctrl+sc_z  areamex
bind           Any+sc_'  togglelos
bind           Any+sc_'  losradar
bind        Ctrl+meta+5  viewta
bind        Ctrl+meta+6  viewspring
bind             meta+5  LastMsgPos
bind             meta+6  ShowPathTraversability
bind             meta+7  ShowMetalMap
bind             meta+8  ShowElevation
bind                f11  luaui selector
bind            Any+f12  screenshot png
bind      Alt+backspace  fullscreen
bind     Ctrl+meta+sc_q  group unset
bind           Alt+sc_q  remove_from_autogroup
bind meta+sc_q,meta+sc_q  drawlabel
bind          meta+sc_q  drawinmap
bind             Any+up  moveforward
bind           Any+down  moveback
bind          Any+right  moveright
bind           Any+left  moveleft
bind         Any+pageup  moveup
bind       Any+pagedown  movedown
bind            Any+alt  movereset
bind            Any+alt  moverotate
bind           Any+ctrl  movetilt
bind          Ctrl+sc_e  select AllMap++_ClearSelection_SelectAll+
bind           Ctrl+tab  select AllMap+_Builder_Idle+_ClearSelection_SelectOne+
bind                tab  select AllMap+_ManualFireUnit_Not_IdMatches_cordecom_Not_IdMatches_armdecom_Not_IdMatches_armthor+_ClearSelection_SelectOne+
bind               sc_q  select Visible+_InPrevSel+_ClearSelection_SelectAll+
bind          Ctrl+sc_q  select PrevSelection++_ClearSelection_SelectPart_50+
bind          Ctrl+sc_w  select AllMap+_InPrevSel+_ClearSelection_SelectAll+
bind            numpad2  moveback
bind            numpad6  moveright
bind            numpad4  moveleft
bind            numpad8  moveforward
bind            numpad9  moveup
bind            numpad3  movedown
bind            numpad1  movefast
bind          backspace  MuteSound
bind            numpad+  snd_volume_increase
bind               sc_=  snd_volume_increase
bind               sc_-  snd_volume_decrease
bind            numpad-  snd_volume_decrease
bind              Alt+0  add_to_autogroup 0
bind                  0  group 0
bind             Ctrl+0  group set 0
bind            Shift+0  group add 0
bind       Ctrl+Shift+0  group selectadd 0
bind         Ctrl+Alt+0  group selecttoggle 0
bind              Alt+1  add_to_autogroup 1
bind                  1  group 1
bind             Ctrl+1  group set 1
bind            Shift+1  group add 1
bind       Ctrl+Shift+1  group selectadd 1
bind         Ctrl+Alt+1  group selecttoggle 1
bind              Alt+2  add_to_autogroup 2
bind                  2  group 2
bind             Ctrl+2  group set 2
bind            Shift+2  group add 2
bind       Ctrl+Shift+2  group selectadd 2
bind         Ctrl+Alt+2  group selecttoggle 2
bind              Alt+3  add_to_autogroup 3
bind                  3  group 3
bind             Ctrl+3  group set 3
bind            Shift+3  group add 3
bind       Ctrl+Shift+3  group selectadd 3
bind         Ctrl+Alt+3  group selecttoggle 3
bind              Alt+4  add_to_autogroup 4
bind                  4  group 4
bind             Ctrl+4  group set 4
bind            Shift+4  group add 4
bind       Ctrl+Shift+4  group selectadd 4
bind         Ctrl+Alt+4  group selecttoggle 4
bind              Alt+5  add_to_autogroup 5
bind                  5  group 5
bind             Ctrl+5  group set 5
bind            Shift+5  group add 5
bind       Ctrl+Shift+5  group selectadd 5
bind         Ctrl+Alt+5  group selecttoggle 5
bind              Alt+6  add_to_autogroup 6
bind                  6  group 6
bind             Ctrl+6  group set 6
bind            Shift+6  group add 6
bind       Ctrl+Shift+6  group selectadd 6
bind         Ctrl+Alt+6  group selecttoggle 6
bind              Alt+7  add_to_autogroup 7
bind                  7  group 7
bind             Ctrl+7  group set 7
bind            Shift+7  group add 7
bind       Ctrl+Shift+7  group selectadd 7
bind         Ctrl+Alt+7  group selecttoggle 7
bind              Alt+8  add_to_autogroup 8
bind                  8  group 8
bind             Ctrl+8  group set 8
bind            Shift+8  group add 8
bind       Ctrl+Shift+8  group selectadd 8
bind         Ctrl+Alt+8  group selecttoggle 8
bind              Alt+9  add_to_autogroup 9
bind                  9  group 9
bind             Ctrl+9  group set 9
bind            Shift+9  group add 9
bind       Ctrl+Shift+9  group selectadd 9
bind         Ctrl+Alt+9  group selecttoggle 9
bind        Ctrl+meta+1  set_camera_anchor 1
bind             meta+1  focus_camera_anchor 1
bind        Ctrl+meta+2  set_camera_anchor 2
bind             meta+2  focus_camera_anchor 2
bind        Ctrl+meta+3  set_camera_anchor 3
bind             meta+3  focus_camera_anchor 3
bind        Ctrl+meta+4  set_camera_anchor 4
bind             meta+4  focus_camera_anchor 4

bind 			Any+sc_a  gridmenu_key 2 1 builder
bind           	Any+sc_s  gridmenu_key 2 2 builder
bind           	Any+sc_d  gridmenu_key 2 3 builder
bind           	Any+sc_w  gridmenu_key 3 2 builder

bind 			sc_w moveforward
bind 			sc_a moveleft
bind 			sc_s moveback
bind 			sc_d moveright
```

Then `/luaui reload` and `/widgetselector` and make sure both the local `Grid Menu` and `Lua UiKeys loader` are enabled.

Result: it WORKS. Camera moves with WASD weeeeeeee. Damn, I love this 😁 Below are some notes (expected / unexpected denotes my personal expectation of how I expected the thing to behave as a user, as if I was toggling an `Enable WASD camera movement` in the settings without any prior knowledge):

- **[Expected]** When no builder nor lab is selected, WASD moves camera.
- **[Expected]** When a builder is selected, WASD moves camera.
- When a builder is selected, and a build category (ZXCV) is selected:
  - **[Expected]** If a build option for WASD is available, WASD selects the build option.
  - **[Unexpected]** If a build option for WASD is NOT available, WASD moves camera.
    - I expected not to be able to move camera after selecting a category, up until I cancel or queue a build order, which removes category selection, or up until I deselect builder.
- When a lab is selected:
  - **[Unexpected]** If a build option for WASD is available, WASD moves camera.
    - I expected WASD to select the build option. (This was already stated as intended for future work in OP.)
  - **[Unexpected]** If a build option for WASD is NOT available, WASD moves camera.
    - I expected not to be able to move camera after selecting a lab, up until I deselect lab.

One funny behaviour:

- If you select a bunch of units and line-drag a move command, then move camera with WASD, the line-drag UI "shadow" does not update but the actual line-drag command does update.

https://user-images.githubusercontent.com/4659919/232344136-0d737bb6-c311-40e5-a43a-6216ce9802f6.mp4

**Comment by thehobojoe** [comment]
This is ready to merge by anyone with permissions

**Pull request merged**